cpp-terminal/platforms/file.cpp

```diff
@@ -3,88 +3,151 @@
 #else
-  m_file = {std::fopen(filename, mode)};
-  if(m_file == nullptr)
+  std::size_t flag{O_ASYNC | O_DSYNC | O_NOCTTY | O_SYNC};
+  if(mode.find('r') != std::string::npos) flag |= O_RDONLY;
+  else if(mode.find('w') != std::string::npos)
+    flag |= O_WRONLY;
+  else
+    flag |= O_RDWR;
+  m_fd = {::open(filename.c_str(), flag)};
+  if(m_fd == -1)
   {
-    m_file   = {std::fopen("/dev/null", mode)};
-    m_isNull = true;
+    m_fd = {::open("/dev/null", flag)};
+    if(m_fd != -1) m_null = true;
+  }
+  if(m_fd != -1)
+  {
+    m_file   = fdopen(m_fd, mode.c_str());
+    m_handle = m_file;
   }
 #endif
+  setvbuf(m_file, nullptr, _IONBF, 0);
 }
 
 Term::Private::FileHandler::~FileHandler()
 {
 #if defined(_WIN32)
-  CloseHandle(reinterpret_cast<HANDLE>(m_file));
-#else
-  std::fclose(reinterpret_cast<FILE*>(m_file));
+  std::fclose(m_file);
 #endif
 }
 
-bool Term::Private::FileHandler::isNull() { return m_isNull; }
+bool Term::Private::FileHandler::null() const { return m_null; }
 
-Term::Private::consoleFileHandler Term::Private::FileHandler::getHandler()
-{
-#if defined(_WIN32)
-  return m_file;
-#else
-  return fileno(reinterpret_cast<FILE*>(m_file));
-#endif
-}
+FILE* Term::Private::FileHandler::file() { return m_file; }
+
+int Term::Private::FileHandler::fd() const { return m_fd; }
+
+Term::Private::FileHandler::Handle Term::Private::FileHandler::handle() { return m_handle; }
 
 int Term::Private::FileInitializer::m_counter = {0};
 
-void Term::Private::FileInitializer::initialize()
+void Term::Private::FileInitializer::init()
 {
   if(m_counter++ == 0)
   {
 #if defined(_WIN32)
-    new(&Term::Private::std_cin) FileHandler("CONIN$", "");
-    new(&Term::Private::std_cout) FileHandler("CONOUT$", "");
+    new(&Term::Private::in) InputFileHandler("CONIN$");
+    new(&Term::Private::out) OutputFileHandler("CONOUT$");
 #else
-    new(&Term::Private::std_cin) FileHandler("/dev/tty", "r");
-    new(&Term::Private::std_cout) FileHandler("/dev/tty", "w");
+    new(&Term::Private::in) InputFileHandler("/dev/tty");
+    new(&Term::Private::out) OutputFileHandler("/dev/tty");
 #endif
   }
 }
 
-Term::Private::FileInitializer::FileInitializer() { initialize(); }
-
 Term::Private::FileInitializer::~FileInitializer()
 {
   if(--m_counter == 0)
   {
-    (&Term::Private::std_cin)->~FileHandler();
-    (&Term::Private::std_cout)->~FileHandler();
+    (&Term::Private::in)->~InputFileHandler();
+    (&Term::Private::out)->~OutputFileHandler();
   }
 }
+
+int Term::Private::OutputFileHandler::write(const std::string& str)
+{
+  if(str.empty()) return 0;
+#if defined(_WIN32)
+  DWORD dwCount{0};
+  if(WriteConsole(handle(), &str[0], str.size(), &dwCount, nullptr) == 0) return -1;
+  else
+    return dwCount;
+#else
+  return ::write(fd(), &str[0], str.size());
+#endif
+}
+
+int Term::Private::OutputFileHandler::write(const char& ch)
+{
+#if defined(_WIN32)
+  DWORD dwCount{0};
+  if(WriteConsole(handle(), &ch, 1, &dwCount, nullptr) == 0) return -1;
```

Comment:
If I see correct the function does not return something when 
`#if defined(_WIN32)` is true
and 
`if(WriteConsole(handle(), &ch, 1, &dwCount, nullptr) == 0)` is false

Reply:
is there a missing `else 0`?

Reply:
Nice catch, I think it should return dwCount